site/en/tutorials/load_data/images.ipynb

```diff
@@ -0,0 +1,1356 @@
+        "\n",
+        "Normally you would specify the real number of steps per epoch, but for demonstration purposes only run 3 steps."
+      ]
+    },
+    {
+      "metadata": {
+        "id": "AnXPRNWoTypI",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "model.fit(ds, epochs=1, steps_per_epoch=3)  # steps_per_epoch=math.ceil(len(all_image_paths)/BATCH_SIZE)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "UMVnoBcG_NlQ",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "## Performance\n",
+        "\n",
+        "Note: This section just shows a couple of easy tricks that may help performance. For an in depth guide see [Input Pipeline Performance](https://www.tensorflow.org/guide/performance/datasets)\n",
+        "\n",
+        "The simple pipeline used above reads each file individually, on each epoch. This is fine for local training on CPU but may not be sufficient for GPU training, and is totally inapprpriate for any sort of distributed training. "
+      ]
+    },
+    {
+      "metadata": {
+        "id": "oNmQqgGhLWie",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "To investigate, first build a simple function to check the performance of our datasets:"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "_gFVe1rp_MYr",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "import time\n",
+        "\n",
+        "def timeit(ds, batches=100):\n",
+        "  overall_start = time.time()\n",
+        "  # Fetch a single batch to prime the pipeline (fill the shuffle buffer),\n",
+        "  # before starting the timer\n",
+        "  it = iter(ds.take(batches+1))\n",
+        "  next(it)\n",
+        "\n",
+        "  start = time.time()\n",
+        "  for i,(images,labels) in enumerate(it):\n",
+        "    if i%10 == 0:\n",
+        "      print('.',end='')\n",
+        "  print()\n",
+        "  end = time.time()\n",
+        "\n",
+        "  duration = end-start\n",
+        "  print(\"{} batches: {} s\".format(batches, duration))\n",
+        "  print(\"{:0.5f} Images/s\".format(BATCH_SIZE*batches/duration))\n",
+        "  print(\"Total time: {}s\".format(end-overall_start))"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "TYiOr4vdLcNX",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The performance of the current dataset is:"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "qwOJGxw_FkJU",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "ds = image_label_ds.repeat()\n",
+        "ds = ds.shuffle(buffer_size=4000) \n",
```

Comment:
OOC, why did you pick 4000? as someone following the tutorial, I would want to know where this magic constant comes from. It might be worth saying that smaller values will result in faster startup (the buffer to initial fill is smaller) while higher values result in better shuffling (in the limit where buffer_size == input_size the shuffling is perfectly random).

Another thing to note is that because you `repeat` before `shuffle` some images might occur for the second time before other images occur at all. This can generally negatively affect accuracy. I suspect the reason you are doing this is to avoid filling up the shuffle buffer at the beginning of each epoch. To avoid doing that, while maintaining clean epoch boundaries, you can either 1) use the [tf.data.experimental.shuffle_and_repeat](https://www.tensorflow.org/api_docs/python/tf/data/experimental/shuffle_and_repeat) transformation or 2) do this:

```
ds = ds.shuffle(...).repeat().
options = tf.data.Options
options.experimental_shuffle_and_repeat_fusion = True
ds = ds.with_options(ds)
```

Reply:
hmmm... the options approach doesn't seem to be working in tf-nightly. I'll send you a bug.

`shuffle_and_repeat` is.
Fixed.

Reply:
The API for this has changed to:

```
ds = ds.shuffle(...).repeat().
options = tf.data.Options
options.experimental_optimization = tf.data.experimental.OptimizationOptions()
options.experimental_optimization.shuffle_and_repeat_fusion = True
ds = ds.with_options(ds)
```